project_euler/problem_074/sol2.py

```diff
@@ -36,18 +40,23 @@ def factorial(a: int) -> int:
     if a < 0:
         raise ValueError("Invalid negative input!", a)
 
+    if a in factorial_cache:
+        return factorial_cache[a]
+
     # The case of 0! is handled separately
     if a == 0:
-        return 1
+        factorial_cache[a] = 1
     else:
         # use a temporary support variable to store the computation
+        temporary_number = a
```

Comment:
Why are you using an extra variable in here?

Reply:
because the number was being reduced by 1 in line 48 and we need the original number for adding an entry to the cache dict